src/EoS/Nuclear/CPU_EoS_Nuclear.cpp

```diff
@@ -454,18 +481,266 @@ static real EoS_DensPres2CSqr_Nuclear( const real Dens_Code, const real Pres_Cod
+         const real Dens2CGS    = AuxArray_Flt[NUC_AUX_DENS2CGS  ];
+         const real Kelvin2MeV  = AuxArray_Flt[NUC_AUX_KELVIN2MEV];
+         const real sEint2Code  = AuxArray_Flt[NUC_AUX_VSQR2CODE ];
+
+         const int  NRho        = AuxArray_Int[NUC_AUX_NRHO ];
+         const int  NEps        = AuxArray_Int[NUC_AUX_NEPS ];
+         const int  NYe         = AuxArray_Int[NUC_AUX_NYE  ];
+         const int  NMode       = AuxArray_Int[NUC_AUX_NMODE];
+
+         real Dens_CGS  = Dens_Code * Dens2CGS;
+         real Temp_MeV  = Temp_Kelv * Kelvin2MeV;
+         real sEint_CGS = NULL_REAL;
+         real Entr      = NULL_REAL;
+         real Useless   = NULL_REAL;
+         int  Err       = NULL_INT;
+
+
+//       check floating-point overflow and Ye
+#        ifdef GAMER_DEBUG
+         if ( Nuc_Overflow(Dens_CGS) )
+            printf( "ERROR : EoS overflow (Dens_CGS %13.7e, Dens_Code %13.7e, Dens2CGS %13.7e) in %s() !!\n",
+                    Dens_CGS, Dens_Code, Dens2CGS, __FUNCTION__ );
+
+         if ( Nuc_Overflow(Temp_MeV) )
+            printf( "ERROR : EoS overflow (Temp_MeV %13.7e, Temp_Kelv %13.7e, Kelvin2MeV %13.7e) in %s() !!\n",
+                    Temp_MeV, Temp_Kelv, Kelvin2MeV, __FUNCTION__ );
+
+         if ( Ye < (real)Table[NUC_TAB_YE][0]  ||  Ye > (real)Table[NUC_TAB_YE][NYe-1] )
+            printf( "ERROR : invalid Ye = %13.7e (min = %13.7e, max = %13.7e) in %s() !!\n",
+                    Ye, Table[NUC_TAB_YE][0], Table[NUC_TAB_YE][NYe-1], __FUNCTION__ );
+#        endif // GAMER_DEBUG
+
+
+//       invoke the nuclear EoS driver
+         nuc_eos_C_short( Dens_CGS, &sEint_CGS, Ye, &Temp_MeV, &Entr, &Useless, &Useless, &Useless,
+                          EnergyShift, NRho, NEps, NYe, NMode,
+                          Table[NUC_TAB_ALL], Table[NUC_TAB_ALL_MODE], Table[NUC_TAB_RHO], Table[NUC_TAB_EPS],
+                          Table[NUC_TAB_YE], Table[NUC_TAB_TEMP_MODE], Table[NUC_TAB_ENTR_MODE], Table[NUC_TAB_PRES_MODE],
+                          Mode, &Err, NULL_REAL );
+
+//       trigger a *hard failure* if the EoS driver fails
+         if ( Err )
+         {
+            sEint_CGS = NAN;
+            Entr      = NAN;
+         }
+
+
+//       store the output results
+         const real Eint_Code = (  ( sEint_CGS + EnergyShift ) * sEint2Code  ) * Dens_Code;
+         Out[0] = Eint_Code;  // volume energy density in code units (with energy shift)
+         Out[1] = Entr;       // entropy in kB/baryon
+
+
+//       final check
+#        ifdef GAMER_DEBUG
+//       still require Eint>0 for the nuclear EoS
+         if ( Hydro_CheckNegative(Eint_Code) )
+         {
+            printf( "ERROR : invalid output internal energy density (%13.7e) in %s() !!\n", Eint_Code, __FUNCTION__ );
+            printf( "        Dens=%13.7e, Temp=%13.7e, Ye=%13.7e\n", Dens_Code, Temp_Kelv, Ye );
+            printf( "        EoS error code: %d\n", Err );
+         }
```

Comment:
It could be better to specify the unit of output quantity here. Same for Entr and those in the entropy mode.